python/tchannel/tornado/dispatch.py

```diff
@@ -92,15 +91,27 @@ def handle_call(self, request, connection):
 
             connection.post_response(response)
 
-            yield self._call_endpoint(
-                endpoint,
-                request,
-                response,
-                TChannelProxy(
-                    connection.tchannel,
-                    request.tracing,
-                ),
-            )
+            try:
+                yield gen.maybe_future(
+                    endpoint(
+                        request,
+                        response,
+                        TChannelProxy(
+                            connection.tchannel,
+                            request.tracing,
+                        ),
+                    )
+                )
+                response.flush()
+            except Exception as e:
+                # refine the exception in the following patches
+                if response.flushed:
+                    # this is bad if user called flush before exception happens
+                    return
+
+                response.write_body(e.message)
```

Comment:
For unexpected exceptions, you want to send an internal error back, not an application error.